sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/ControlFlowGraph/UcfgExpression.cs

```diff
@@ -18,23 +18,191 @@
  * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
  */
 
+using System;
 using Microsoft.CodeAnalysis;
+using SonarAnalyzer.Helpers;
 using SonarAnalyzer.Protobuf.Ucfg;
 
 namespace SonarAnalyzer.ControlFlowGraph.CSharp
 {
-    internal static class UcfgExpression
+    internal abstract class UcfgExpression
     {
-        public static readonly Expression Constant =
-            new Expression { Const = new Constant { Value = "\"\"" } };
+        public static UcfgExpression This { get; } = new ThisExpression();
+        public static UcfgExpression Constant { get; } = new ConstantExpression();
+        public static UcfgExpression Unknown { get; } = new UnknownExpression();
 
-        public static readonly Expression This =
-            new Expression { This = new This() };
+        protected UcfgExpression(ISymbol symbol)
+        {
+            TypeSymbol = symbol.GetSymbolType();
+        }
 
-        public static Expression FromVariable(string variableName) =>
-            new Expression { Var = new Variable { Name = variableName } };
+        public abstract Expression Expression { get; }
+        public ITypeSymbol TypeSymbol { get; }
 
-        public static Expression FromNamedType(INamedTypeSymbol namedType) =>
-            new Expression { Classname = new ClassName { Classname = UcfgIdentifier.CreateTypeId(namedType).ToString() } };
+        public virtual void ApplyAsTarget(Instruction instruction) =>
+            throw new UcfgBusinessException($"Type {GetType().Name} could not be applied as target of the instruction.");
+
+        internal class UnknownExpression : ConstantExpression
```

Comment:
?Could be private? The parent class will still be able to construct it.
Not sure if the other sub-classes could be _private_ too or just _sealed_.